File: java/exercises/01-beginner/ex016_big_decimal_money/BigDecimalMoneyTest.java
package fewolearning.exercises.beginner.ex016_big_decimal_money;

import org.junit.jupiter.api.Test;

import java.math.BigDecimal;
import java.math.RoundingMode;

import static org.junit.jupiter.api.Assertions.assertEquals;

class BigDecimalMoneyTest {

    @Test
    void addSumsTwoAmountsExactly() {
        BigDecimal sum = BigDecimalMoney.add(new BigDecimal("1.10"), new BigDecimal("2.20"));

        assertEquals(0, new BigDecimal("3.30").compareTo(sum));
    }

    @Test
    void roundToCentsRoundsHalfUp() {
        BigDecimal rounded = BigDecimalMoney.roundToCents(new BigDecimal("1.005"), RoundingMode.HALF_UP);

        assertEquals(0, new BigDecimal("1.01").compareTo(rounded));
        assertEquals(2, rounded.scale());
    }

    @Test
    void roundToCentsRoundsDownWhenAskedTo() {
        BigDecimal rounded = BigDecimalMoney.roundToCents(new BigDecimal("1.999"), RoundingMode.DOWN);

        assertEquals(0, new BigDecimal("1.99").compareTo(rounded));
        assertEquals(2, rounded.scale());
    }
}

File: kotlin/exercises/01-beginner/ex031_big_decimal_money/BigDecimalMoneyTest.kt
package fewolearning.exercises.beginner.ex031_big_decimal_money

import java.math.BigDecimal
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.Assertions.assertEquals

class BigDecimalMoneyTest {

    @Test
    fun addSumsTwoMonetaryAmountsExactly() {
        val result = add(BigDecimal("1.50"), BigDecimal("2.30"))

        assertEquals(0, result.compareTo(BigDecimal("3.80")))
    }

    @Test
    fun roundToCentsRoundsHalfUpToTwoDecimalPlaces() {
        assertEquals(0, roundToCents(BigDecimal("2.345")).compareTo(BigDecimal("2.35")))
        assertEquals(0, roundToCents(BigDecimal("2.344")).compareTo(BigDecimal("2.34")))
    }

    @Test
    fun roundToCentsPadsAWholeAmountToTwoDecimalPlaces() {
        val result = roundToCents(BigDecimal("5"))

        assertEquals(0, result.compareTo(BigDecimal("5.00")))
        assertEquals(2, result.scale())
    }
}
